api/src/server/websocket.rs

```diff
@@ -215,18 +221,29 @@ impl StreamHandler<ws::Message, ws::ProtocolError> for ProgramWebsocket {
     }
 }
 
-/// Do websocket handshake, look up the request Environment by ID, then (if it
+/// Do websocket handshake, look up the request ProgramSpec by ID, then (if it
 /// exists), start a handler for it.
-pub fn ws_environments_by_id(
+pub fn ws_program_specs_by_id(
     r: HttpRequest,
     pool: web::Data<Pool>,
-    env_id: web::Path<i32>,
+    program_spec_id: web::Path<i32>,
     stream: web::Payload,
 ) -> Result<HttpResponse, actix_web::Error> {
     let conn: &PgConnection = &pool.get().unwrap();
-    let env = environments::dsl::environments
-        .find(env_id.into_inner())
+    let program_spec: ProgramSpec = program_specs::dsl::program_specs
+        .find(program_spec_id.into_inner())
         .get_result(conn)
         .map_err(ServerError::from)?;
-    ws::start(ProgramWebsocket::new(env), &r, stream)
+
+    // As far as I can tell, Diesel doesn't have an easier way to do a foreign
+    // key lookup
+    let hardware_spec: HardwareSpec = hardware_specs::dsl::hardware_specs
+        .find(program_spec.hardware_spec_id)
+        .get_result(conn)
+        .map_err(ServerError::from)?;
+    ws::start(
+        ProgramWebsocket::new(hardware_spec, program_spec),
+        &r,
+        stream,
```

Comment:
Could you do a join so you can do it in one query?

Reply:
Yeah so normally I would've but the Diesel docs actually recommend using separate queries:
> Associated data is typically loaded in multiple queries (one query per table). This is usually more efficient than using a join, especially if 3 or more tables are involved. For most datasets, using a join to load in a single query transmits so much duplicate data that it costs more time than the extra round trip would have.
As I'm typing this tho, I'm realizing that might only refer to one-to-many relationships, not when you're just fetching a single related row. I'll look into doing a join, and if it's not too hard I'll do that.
https://docs.rs/diesel/1.4.3/diesel/associations/index.html